algorithms/indexing/indexer.py

```diff
@@ -412,9 +412,10 @@ def from_parameters(
                 from dxtbx.imageset import ImageSet  # , MemImageSet
 
                 for experiment in experiments:
-                    experiment.imageset = ImageSet(
-                        experiment.imageset.data(), experiment.imageset.indices()
-                    )
+                    if isinstance(experiment.imageset, ImageSequence):
+                        experiment.imageset = ImageSet(
+                            experiment.imageset.data(), experiment.imageset.indices()
+                        )
```

Comment:
It isn't clear to me why indexing is even creating imagesets in the first place... it appears to only be doing this in the stills indexer, so presumably you might have a better idea why this is the case?

Reply:
Right, so in `dials.stills_process`, [imageset_from_anyset](https://github.com/cctbx/dxtbx/blob/main/imageset.py#L564) is [called](https://github.com/dials/dials/blob/main/command_line/stills_process.py#L297) on all images right at the start.  So by the time the indexing is ran, there's no way to have an ImageSequence.

However, if `dials.index` is ran, with `indexing.stills.indexer=stills`, then it's possible an ImageSequence makes it this far, even though the user wishes it to be treated as a still.  Therefore, the ImageSequence is reset as an ImageSet, and things like the StillsReflectionPredictor are [properly picked](https://github.com/dials/dials/blob/main/algorithms/spot_prediction/reflection_predictor.py#L88).

In the `dials.stills_process` case, before this changeset, ImageSets were always recreated, which invalidated the cache and caused a data read later on.  With this changeset, they are never recreated, because of the imageset_from_anyset call.

A data read wouldn't be triggered in the `dials.index` case because the data is never read in that program, so the cache doesn't matter.

Reply:
I see. It seems a bit odd that refinement should care at all about the imageset - it _should_ only depend on the models and input parameters. Perhaps a better (at least longer term) solution would be for refinement not to depend on the imageset type when choosing which predictor to use, then constructing a new imageset in indexing shouldn't be necessary at all.

Reply:
Totally agree.  My googling indicates that picking the StillsReflectionPredictor is the only place where ImageSequence is tested for explicitly, and I think it should instead be choosing the predictor based on if the scan is a set of stills:

Instead of 

```
if isinstance(experiment.imageset, ImageSequence):
```

It should be

```
if experiment.scan and not experiment.scan.is_still():
```

Like it is elsewhere.

But that seemed like a bigger change than this fix warranted?  Maybe not?  If folks like that better I can add it to this PR.

Reply:
TBH I think this is not a question of "like or dislike" - I think very shortly this will become necessary for sane processing of grid scan data beyond simply spot finding. As such, I'd probably suggest we put together a second PR where we actively eliminate 

```
if isinstance(experiment.imageset, ImageSequence):
```

type checks from the code base - this would be work which was started back in the load before heat death days but never (yet) completed.

Reply:
On a practical note @phyy-nx  if you could put a comment in there as to _why_ this check is in place would probably help for the medium term.